### ISSUE TITLE
New Stochastic Optimization Algorithms in Pytorch

### ISSUE THREAD

**Issue by touqir14** [pytorch/pytorch#28194]
## 🚀 Feature
I would like to suggest new stochastic optimizer additions to Pytorch.

### For non-convex loss functions
It is known that adaptive stochastic optimizers like Adam, Adagrad, RMSprop can indeed fail to converge. Moreover the test error of the trained models can be larger when compared to the models trained by SGD even though if they attain lower training error than with SGD - in other words they can overfit. See [[1](http://papers.nips.cc/paper/7003-the-marginal-value-of-adaptive-gradient-methods-in-machine-learning.pdf)].

Recently, ADABOUND was proposed [[2](https://openreview.net/pdf?id=Bkg3g2R9FX)] to overcome the above mentioned problems of the adaptive optimizers. The algorithm combines the best of both worlds : (a) Makes fast progress initially like the adaptive methods, and (b) attains similar or better test accuracy to that of SGD. In terms of code size and complexity, it is close to Adam's.

### For convex loss functions
While Pytorch is generally used for deep learning purposes, I see it is as a tensor-based Machine Leaning (ML) library with automatic differentiation that can run on both CPU and GPU and which allows easy implementation of both convex and non-convex ML models. As such, I think an ML engineer working on convex ML models would be interested to see faster state-of-the-art stochastic optimizers implemented in Pytorch like the variance reduced optimizers. Among them, SVRG [[3](https://papers.nips.cc/paper/4937-accelerating-stochastic-gradient-descent-using-predictive-variance-reduction.pdf)] is most widely known. In terms of code size and complexity, it s somewhat close to SGD's with few modifications.

### Alternatives
I recommended SVRG since it is the most popular variance reduced algorithm. There are more complex variance reduced algorithms that are faster than SVRG and if they are of interest I can suggest some. Also, I would like to know if there are interests on batch optimizers other than LBFGS which is within Pytorch. For large scale ML models, there are cases when LBFGS requires more memory than the system can offer. An alternative would to be try out the accelerated gradient descent that consumes the same amount of memory as Batch gradient descent while performing significantly faster than it. Accelerated gradient descent's code size and complexity is close to batch gradient descent and as such it is way easier to maintain than LBFGS's. Lastly, if there are interests in other specific optimizers, I would be interested to hear out. I can try implementing them.

## References
[1] http://papers.nips.cc/paper/7003-the-marginal-value-of-adaptive-gradient-methods-in-machine-learning.pdf
[2] https://openreview.net/pdf?id=Bkg3g2R9FX
[3] https://papers.nips.cc/paper/4937-accelerating-stochastic-gradient-descent-using-predictive-variance-reduction.pdf

**Comment by harshitsankhla** [comment]
Hi,

I am interested in implementing 3. Let me know if we can work together.

**Comment by 1e100** [comment]
FWIW I was not able to get the official implementation of AdaBound to perform any better than plain SGD on object detection (Retinanet) and semantic segmentation (Deeplab v3+). Could be a case of "holding it wrong" for sure, but I've tried all sorts of training regimes before I gave up on it. I'd be curious to hear of any successes with it, meaning any cases where it reached or exceeded the SGD result on the same architecture, possibly in a shorter time. I'm not super interested in cases where I have to sacrifice mAP/mIoU for training speed. It does intuitively seem, based on my reading of the paper, that it should provide benefit.
